src/internal/connector/onedrive/collection.go

```diff
@@ -129,37 +142,92 @@ func (oc *Collection) populateItems(ctx context.Context) {
 	defer colCloser()
 	defer close(folderProgress)
 
+	limitCh := make(chan struct{}, urlPrefetchChannelBufferSize)
+	defer close(limitCh)
+
+	type uerr struct {
+		itemID string
+		err    error
+	}
+
+	errCh := make(chan uerr, len(oc.driveItemIDs))
+	defer close(errCh)
+
+	ffail := int64(0) // handling fast-fail
 	for _, itemID := range oc.driveItemIDs {
-		// Read the item
-		itemInfo, itemData, err := oc.itemReader(ctx, oc.service, oc.driveID, itemID)
-		if err != nil {
-			errs = support.WrapAndAppendf(itemID, err, errs)
+		if ffail > 0 {
+			break
+		}
+		limitCh <- struct{}{}
+
+		wg.Add(1)
+
+		go func(itemID string) {
+			defer func() { <-limitCh }()
```

Comment:
Due to the goroutine there's a possibility of a panic here if limitCh has been closed as a result of early exit on fast-fail.  A panic recovery would be helpful.  You can additionally protect against this by setting `limitCh` to `nil` on close, and having this defer first check `if limitCh != nil`.

Reply:
Since we have a `wg.Wait()` below before we close, I don't think we will run into this. That said, I have flipped the order of the defer statement so that `wg.Done()` happens after reading from the channel.